File: Problems/SubarraysWithKDifferentIntegers/SubarraysWithKDifferentIntegers.py
'''
Can solve by using "numberOfSubarraysWithAtMostK(k) - numberOfSubarraysWithAtMostK(k - 1)"

This is the most straightforward, because with exactly K distint, it is not as easy to add all of the subarrays
'''

class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        atMostK = self.subarraysWithAtMostKDistinct(nums, k)
        atMostKMinusOne = self.subarraysWithAtMostKDistinct(nums, k - 1)
        return atMostK - atMostKMinusOne
        
    def subarraysWithAtMostKDistinct(self, nums: List[int], k: int) -> int:
        windowStart = 0
        charFrequencyMap = {}
        numSubarrays = 0
        
        for windowEnd in range(len(nums)):
            rightNum = nums[windowEnd]
            
            charFrequencyMap[rightNum] = charFrequencyMap.get(rightNum, 0) + 1
            
            while len(charFrequencyMap) > k:
                leftNum = nums[windowStart]
                
                charFrequencyMap[leftNum] -= 1
                
                if charFrequencyMap[leftNum] == 0:
                    del charFrequencyMap[leftNum]
                    
                windowStart += 1
            
            # Add the number of subarrays ending at windowEnd (equal to size of window)
            numSubarrays += windowEnd - windowStart + 1
        
        return numSubarrays
        
        

File: Problems/SubarraysWithKDifferentIntegers/SubarraysWithKDifferentIntegers.cpp
class Solution {
public:
    int subarraysWithKDistinct(vector<int>& nums, int k) {
        int atMostK = atMostKDistinct(nums, k);
        int atMostKMinusOne = atMostKDistinct(nums, k - 1);
        //cout << "atMostK: " << atMostK << "  atMostKMinusOne: " << atMostKMinusOne << "\n";
        return atMostK - atMostKMinusOne;
    }
    
    // Returns Number of Subarrays with at most K Distint Integers
    int atMostKDistinct(vector<int>& nums, int k) {
        int left = 0;
        int numSubarrays = 0;
        unordered_map<int, int> freqMap;
        
        for(int right = 0; right < nums.size(); right++) {
            int rightNum = nums[right];
            freqMap[rightNum]++;
            
            while(freqMap.size() > k) {
                int leftNum = nums[left];
                freqMap[leftNum]--;
                if(freqMap[leftNum] == 0) {
                    freqMap.erase(leftNum);
                }
                left++;
            }
            
            // Add all subarrays ending at right
            numSubarrays += (right - left + 1);
        }
        return numSubarrays;
        
    }
};
